pkg/ccl/sqlproxyccl/proxy.go

```diff
@@ -108,15 +109,32 @@ func (s *Server) Proxy(conn net.Conn) error {
 		return newErrorf(CodeUnexpectedStartupMessage, "unsupported post-TLS startup message: %T", m)
 	}
 
-	outgoingAddr, outgoingTLS, clientErr := s.opts.BackendFromParams(msg.Parameters)
+	backendConfig, clientErr := s.opts.BackendConfigFromParams(msg.Parameters)
```

Comment:
Rather than returning an admitter to check in `backendConfig.Admitter`, could you instead pass `ipAddress` as an additional argument to `BackendConfigFromParams` and do the IP-based rate limiting within that function itself?

Reply:
Sure, should we also hide the caching of known connections in this function as well then? Perhaps `BackendConfigFromParams` can additionally return a callback stored in `BackendConfig`, say `func CacheConn() error`, to be called after successful login? This will replace the need to provide a `ClientID` that is cached by the proxy.